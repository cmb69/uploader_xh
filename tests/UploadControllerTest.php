<?php

/**
 * Copyright 2023 Christoph M. Becker
 *
 * This file is part of Uploader_XH.
 *
 * Uploader_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Uploader_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with Uploader_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Uploader;

use ApprovalTests\Approvals;
use PHPUnit\Framework\TestCase;
use Plib\FakeRequest;
use Plib\Jquery;
use Plib\View;

class UploadControllerTest extends TestCase
{
    /** @var UploadController */
    private $sut;

    /** @var Jquery&MockObject */
    private $jquery;

    /** @var FileSystemService&MockObject */
    private $fileSystemService;

    public function setUp(): void
    {
        $plugin_cf = XH_includeVar("./config/config.php", 'plugin_cf');
        $conf = $plugin_cf['uploader'];
        $plugin_tx = XH_includeVar("./languages/en.php", 'plugin_tx');
        $lang = $plugin_tx['uploader'];
        $fileFolders = [
            'images' => 'irrelevant_images',
            'downloads' => 'irrelevant_downloads',
            'media' => 'irrelevant_media',
            'userfiles' => 'irrelevant_userfiles',
        ];
        $this->jquery = $this->createStub(Jquery::class);
        $this->fileSystemService = $this->createStub(FileSystemService::class);
        $this->sut = new UploadController(
            1,
            $conf,
            "./",
            $fileFolders,
            $this->jquery,
            $this->fileSystemService,
            "2M",
            new View("./views/", $lang)
        );
    }

    public function testDefaultActionRendersPlaceholder(): void
    {
        $response = ($this->sut)(new FakeRequest(), null, null, null);
        Approvals::verifyHtml($response->output());
    }

    public function testWidgetActionRendersWidget(): void
    {
        $this->fileSystemService->method('getSubdirsOf')->willReturn(["/"]);
        $request = new FakeRequest(["url" => "http://example.com/?&uploader_serial=1"]);
        $response = ($this->sut)($request, null, null, null);
        Approvals::verifyHtml($response->output());
    }

    public function testIgnoresUnrelatedRequests(): void
    {
        $this->fileSystemService->method('getSubdirsOf')->willReturn(["/"]);
        $request = new FakeRequest(["url" => "http://example.com/?&uploader_serial=17"]);
        $response = ($this->sut)($request, null, null, null);
        $this->assertSame("", $response->output());
    }

    public function testRendersDynamicWidget(): void
    {
        $this->fileSystemService->method('getSubdirsOf')->willReturn(["/foo"]);
        $this->fileSystemService->method("isDir")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?&uploader_type=downloads&uploader_subdir=%2Ffoo&uploader_serial=1",
        ]);
        $response = ($this->sut)($request, null, null, null);
        Approvals::verifyHtml($response->output());
    }
}
