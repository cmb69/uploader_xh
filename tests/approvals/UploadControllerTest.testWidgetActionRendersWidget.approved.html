
<div class="uploader_widget" data-config='{"url":"/?&uploader_action=upload&uploader_serial=1&uploader_type=images&uploader_subdir=%2F&uploader_resize","filters":{"max_file_size":"20000000b","mime_types":[{"title":"Images","extensions":"jpg,gif,png"}]},"flash_swf_url":"./lib/Moxie.swf","silverlight_xap_url":"./lib/Moxie.xap","file_data_name":"uploader_file","chunk_size":"2mb"}'>
  <div class="uploader_controls">
    <select class="uploader_type" title="Type" data-url="/?&amp;uploader_action=widget&amp;uploader_serial=1&amp;uploader_type=FIXME&amp;uploader_subdir=%2F&amp;uploader_resize">
      <option value="images" selected>images</option>
      <option value="downloads" >downloads</option>
      <option value="media" >media</option>
      <option value="userfiles" >userfiles</option>
    </select>
    <select class="uploader_subdir" title="Sub-Folder" data-url="/?&amp;uploader_action=widget&amp;uploader_serial=1&amp;uploader_type=images&amp;uploader_subdir=FIXME&amp;uploader_resize">
      <option value="/" selected>/</option>
    </select>
    <select class="uploader_resize" title="Resize" data-url="/?&amp;uploader_action=widget&amp;uploader_serial=1&amp;uploader_type=images&amp;uploader_subdir=%2F&amp;uploader_resize=FIXME">
      <option value="" selected></option>
      <option value="small" >small</option>
      <option value="medium" >medium</option>
      <option value="large" >large</option>
    </select>
  </div>
  <table class="uploader_filelist">
    <tr>
      <th class="uploader_filename">Filename</th>
      <th class="uploader_size">Size</th>
      <th class="uploader_progress">State</th>
      <th></th>
    </tr>
    <tr class="uploader_row_template">
      <td class="uploader_filename"></td>
      <td class="uploader_size"></td>
      <td class="uploader_progress"></td>
      <td><button class="uploader_remove">Remove</button></td>
    </tr>
  </table>
  <div class="uploader_buttons">
    <button class="uploader_pickfiles">Select files</button>
    <button class="uploader_uploadfiles">Upload files</button>
  </div>
</div>
